### ISSUE TITLE
Tasks hang when operating on writing Zarr-backed Dataset

### ISSUE THREAD

**Issue by ljstrnadiii** [pydata/xarray#7952]
### What happened?

When writing a dataset to zarr, we sometimes see that the last few tasks hang indefinitely with no cpu activity or data transfer activity (as seen from the daskui). Inspecting the daskui always shows we are waiting on a task with a name like ('store-map-34659153bd4dc964b4e5f380dacebdbe', 0, 1).

### What did you expect to happen?

For all tasks to finish __and__ all taking approximately the same amount of time to complete. I also would expect worker saturation to have some effect and kick in a queue with xarray and map_blocks to zarr, but I don't see this behavior. I do see this behavior with the dask counterpart (example in code the extra section below).

### Minimal Complete Verifiable Example

```Python
import os
import xarray as xr
import dask.array as da
from distributed import Client


def main():
    client = Client("...")

    # task 1: make dataset
    variables = [str(i) for i in range(36)]
    dset = xr.Dataset(
        {
            v: (
                ("x", "y"),
                da.ones((50000, 50000), chunks=(4096, 4096), dtype="float32"),
            )
            for v in variables
        },
        coords={"x": range(50000), "y": range(50000)},
    )

    bucket = "..."
    path = f"gs://{bucket}/tmp/test/0"
    dset.to_zarr(os.path.join(path, "1"))

    # task 2: "preprocess" the dataset with map_blocks
    rt = xr.open_zarr(os.path.join(path, "1"))

    def f(block):
        # I have found copying helps reproduce the issue
        # and is often happening anyways within functions we apply
        # with map blocks e.g. .rio.clip, .fillna, etc.
        block = block.copy().copy()
        return block

    # first setup a target store
    target_path = os.path.join(path, "2")
    rt.to_zarr(target_path, compute=False)

    # I could not reproduce without writing a region
    region = {"x": slice(0, 20000), "y": slice(0, 10000)}
    subset = rt.isel(**region)
    subset.map_blocks(f, template=subset).to_zarr(target_path, region=region)


if __name__ == "__main__":
    main()

```

### MVCE confirmation

- [X] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [X] Complete example — the example is self-contained, including all data and the text of any traceback.
- [X] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [X] New issue — a search of GitHub Issues suggests this is not a duplicate.

### Anything else we need to know?

My dask cluster is deployed with helm with (32) workers each with 1cpu and 4gb.

We often mitigate this issue (poorly) using environment variables to trigger worker restarts so long hanging tasks get rescheduled so the program can complete, but at the cost of the expected time to restart, which can slow down performance.
```        
"DASK_DISTRIBUTED__WORKER__LIFETIME__DURATION": "8 minutes",
"DASK_DISTRIBUTED__WORKER__LIFETIME__STAGGER": "4 minutes",
"DASK_DISTRIBUTED__WORKER__LIFETIME__RESTART": "True",
```

In the mre above, we write `to_zarr` with region often because we often have to throttle the number of tasks submitted to dask.

Possibly relevant links:
- https://github.com/dask/distributed/issues/391
- https://github.com/fsspec/gcsfs/issues/379
- https://github.com/pydata/xarray/issues/4406

__note__: I have tried to minimize the example even more with
```
array = da.random.random((250000, 250000), chunks=(4096, 4096)).astype("float32")
def f(block):
    block = block.copy().copy()
    return block
array.map_blocks(f).to_zarr("gs://...")
```
but I am not seeing the issue. I am seeing worker saturation taking some effect, which might be throttling file operations to gcs causing me to wonder why `map_blocks().to_zarr()` does not see the worker saturation effects and if that would help alleviate issues `gcsfs` is probably seeing.

### Environment

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.10.11 | packaged by conda-forge | (main, May 10 2023, 18:58:44) [GCC 11.3.0]
python-bits: 64
OS: Linux
OS-release: 5.15.0-1036-gcp
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: C.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.12.2
libnetcdf: None

xarray: 2023.5.0
pandas: 1.5.3
numpy: 1.23.5
scipy: 1.10.1
netCDF4: None
pydap: None
h5netcdf: None
h5py: 3.8.0
Nio: None
zarr: 2.15.0
cftime: None
nc_time_axis: None
PseudoNetCDF: None
iris: None
bottleneck: None
dask: 2023.6.0
distributed: 2023.6.0
matplotlib: 3.7.1
cartopy: None
seaborn: 0.12.2
numbagg: None
fsspec: 2023.6.0 # and gcsfs.__version__ == '2023.6.0'
cupy: None
pint: None
sparse: None
flox: None
numpy_groupies: None
setuptools: 67.7.2
pip: 22.3.1
conda: None
pytest: 7.3.2
mypy: 1.3.0
IPython: 8.14.0
sphinx: None

</details>

**Comment by dcherian** [comment]
We'll need some help from someone on the dask/distributed team it seems.

cc @fjetter

**Comment by fjetter** [comment]
One thing I noticed right away in the "data generation task1", is that the following code runs for about ~4min

```python
path = "s3://..."
write_dataset = dset.to_zarr(path, compute=False)
```

and is already writing all the metadata synchronously, i.e. my local client is writing one file after the other. Is this intentional? Computing, i.e. generating and writing the data only takes a fraction of time for me (and is successful)
(Note: As said, I ran this on AWS using a much larger coiled cluster of 50 m6i.large instances, i.e. that's 50x (8GiB + 2CPUs))

This is not what this issue is about but my sense is that you'll want this to be done in a dask task and not done synchronously.

---

About the reported problem, I could not reproduce this on AWS + Coiled.

If this is stuck in a `store-map-XYZ` task my suspicion is that the GCP backend is somehow missing an internal timeout or smth like that.

To diagnose this, I recommend opening the dask dashboard while you are in this stuck state and enter the Info panel (on the right)
![image](https://github.com/pydata/xarray/assets/8629629/cd12dc71-ce59-4e64-9c4b-38269227e447)

This will show a list of workers like this 
![image](https://github.com/pydata/xarray/assets/8629629/7e61f9fa-b5bc-4a17-8f76-b20601040a17)


You will likely have many In-memory tasks everywhere. You should have only one Worker with one task in `Processing`. This is the worker that is stuck. Click on the worker name/address on the left side.

This will look roughly like this
![image](https://github.com/pydata/xarray/assets/8629629/6cbed30d-a41b-42af-a11a-16a2bf73f02b)

The important bit is the righthand side with the one task processing. Click the green button `Call stacks`. which should show you a stack similar to

```python
Key: ('f-5c61e17f47b2a74f5e7444675c1f5dc7', 0, 0)
File "/opt/coiled/env/lib/python3.10/threading.py", line 973, in _bootstrap self._bootstrap_inner()

File "/opt/coiled/env/lib/python3.10/threading.py", line 1016, in _bootstrap_inner self.run()

File "/opt/coiled/env/lib/python3.10/threading.py", line 953, in run self._target(*self._args, **self._kwargs)

File "/opt/coiled/env/lib/python3.10/site-packages/distributed/threadpoolexecutor.py", line 57, in _worker task.run()

File "/opt/coiled/env/lib/python3.10/site-packages/distributed/_concurrent_futures_thread.py", line 65, in run result = self.fn(*self.args, **self.kwargs)

File "/opt/coiled/env/lib/python3.10/site-packages/distributed/utils.py", line 1517, in <lambda> executor, lambda: context.run(func, *args, **kwargs)

File "/opt/coiled/env/lib/python3.10/site-packages/distributed/worker.py", line 3053, in apply_function msg = apply_function_simple(function, args, kwargs, time_delay)

File "/opt/coiled/env/lib/python3.10/site-packages/distributed/worker.py", line 3089, in apply_function_simple result = function(*args, **kwargs)

File "/opt/coiled/env/lib/python3.10/site-packages/distributed/worker.py", line 2944, in execute_task return func(*map(execute_task, args))

File "/Users/fjetter/miniforge3/envs/coiled-benchmarks/lib/python3.10/site-packages/xarray/core/parallel.py", line 268, in _wrapper

File "/var/folders/h0/kd1gdptx7gzb6cbywplfrx1r0000gn/T/ipykernel_81835/1759722540.py", line 5, in f
```

This shows you what this task is currently doing. In my example, I just added a sleep in `f` but I suspect you will see a GCP API call at the very bottom of this stack.

Restarting the dask workers fixes this because it just reschedules the task again s.t. this API request is attempted again and is not timing out. 
If this is all true, this is where your problem is and you need to open an issue with the GCP lib.

**Comment by ljstrnadiii** [comment]
I am no longer able to reproduce 😢. I looked at monitoring last week and compared to today, but traffic/requests/etc all seem to be roughly the same. And I was testing in my personal temp bucket. I might be encountering some throttling or rate request limits on gcp side. 
```
File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/threading.py", line 973, in _bootstrap self._bootstrap_inner()

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/threading.py", line 1016, in _bootstrap_inner self.run()

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/threading.py", line 953, in run self._target(*self._args, **self._kwargs)

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/site-packages/distributed/threadpoolexecutor.py", line 57, in _worker task.run()

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/site-packages/distributed/_concurrent_futures_thread.py", line 65, in run result = self.fn(*self.args, **self.kwargs)

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/site-packages/distributed/utils.py", line 1471, in <lambda> executor, lambda: context.run(func, *args, **kwargs)

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/site-packages/distributed/worker.py", line 3035, in apply_function msg = apply_function_simple(function, args, kwargs, time_delay)

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/site-packages/distributed/worker.py", line 3071, in apply_function_simple result = function(*args, **kwargs)

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/site-packages/distributed/worker.py", line 2926, in execute_task return func(*map(execute_task, args))

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/site-packages/dask/array/core.py", line 4398, in store_chunk return load_store_chunk(x, out, index, lock, return_stored, False)

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/site-packages/dask/array/core.py", line 4380, in load_store_chunk out[index] = x

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/site-packages/zarr/core.py", line 1386, in __setitem__ self.set_orthogonal_selection(pure_selection, value, fields=fields)

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/site-packages/zarr/core.py", line 1573, in set_orthogonal_selection self._set_selection(indexer, value, fields=fields)

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/site-packages/zarr/core.py", line 1859, in _set_selection self._chunk_setitems(lchunk_coords, lchunk_selection, chunk_values,

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/site-packages/zarr/core.py", line 2053, in _chunk_setitems cdatas = {key: self._process_for_setitem(key, sel, val, fields=fields)

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/site-packages/zarr/core.py", line 2053, in <dictcomp> cdatas = {key: self._process_for_setitem(key, sel, val, fields=fields)

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/site-packages/zarr/core.py", line 2145, in _process_for_setitem cdata = self.chunk_store[ckey]

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/site-packages/zarr/storage.py", line 1397, in __getitem__ return self.map[key]

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/site-packages/fsspec/mapping.py", line 143, in __getitem__ result = self.fs.cat(k)

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/site-packages/fsspec/asyn.py", line 121, in wrapper return sync(self.loop, func, *args, **kwargs)

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/site-packages/fsspec/asyn.py", line 94, in sync if event.wait(1):

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/threading.py", line 607, in wait signaled = self._cond.wait(timeout)

File "/usr/mambaforge/envs/redwood-dev/lib/python3.10/threading.py", line 324, in wait gotit = waiter.acquire(True, timeout)
```

**Comment by ljstrnadiii** [comment]
@dcherian or @fjetter any intuition why worker saturation has an effect for a single dask array with `to_zarr` and not an xarray dataset? If this is a gcs rate issue, that might be a nice way to help throttle things. 

I suppose we can close this and I can figure out how to articulate this issue with gcsfs?

**Comment by ljstrnadiii** [comment]
Other suspiciously similar issues:
- https://github.com/fsspec/filesystem_spec/issues/1136
- https://github.com/fsspec/filesystem_spec/issues/565
